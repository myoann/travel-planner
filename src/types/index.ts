type TLanguage = {
  code: string;
  name: string;
  native: string;
  rtl: boolean;
};

type TSubdivision = {
  code: string;
  emoji: string;
  name: string;
};

type TState = {
  code: string;
};

type TContinent = {
  code: string;
  countries: {
    code: string;
  }[];
  name: string;
};

export type TCountry = {
  awsRegion: string;
  capital: string;
  code: string;
  continent: TContinent;
  currencies: string[];
  currency: string;
  emoji: string;
  emojiU: string;
  languages: TLanguage[];
  name: string;
  native: string;
  phone: string;
  phones: string[];
  states: TState[];
  subdivisions: TSubdivision[];
};

export type TOption = {
  key: string;
  name: string;
  value: string;
};
