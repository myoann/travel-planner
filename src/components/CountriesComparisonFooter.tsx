"use client";

import Image from "next/image";

import { TCountry } from "@/types";

type TProps = {
  selectedCountries: TCountry[];
  removeCountry: (country: TCountry) => void;
};

function RemoveButton({ onClick }: { onClick: () => void }) {
  return (
    <button
      onClick={onClick}
      className="absolute right-0 top-0 rounded-full bg-white"
      style={{ transform: "translate(50%, -50%)" }}
    >
      <Image src="cross.svg" alt="Remove image" width={16} height={16} />
    </button>
  );
}

export default function CountriesComparisonFooter({
  selectedCountries,
  removeCountry,
}: TProps) {
  return (
    <div className="fixed bottom-0 z-50 w-full bg-white shadow-2xl">
      <p className="flex h-10 w-full items-center justify-center bg-[#1d1d1f] font-bold text-white">
        Compare countries ({selectedCountries.length}/2)
      </p>

      <div className="flex h-32 items-center justify-center bg-white sm:h-52 sm:flex-col sm:gap-4 md:flex-row md:gap-16">
        <div className="flex text-center">
          <div className="flex w-32 flex-col items-center">
            <div className="relative">
              <Image
                src={`/flags/${selectedCountries[0].code.toLowerCase()}.svg`}
                alt={selectedCountries[0].name}
                width={60}
                height={40}
              />

              <RemoveButton
                onClick={() => removeCountry(selectedCountries[0])}
              />
            </div>

            <p>{selectedCountries[0].name}</p>
          </div>

          <span className="mx-4 self-center text-4xl">+</span>

          {selectedCountries.length > 1 ? (
            <div className="flex w-32 flex-col items-center">
              <div className="relative">
                <Image
                  src={`/flags/${selectedCountries[1].code.toLowerCase()}.svg`}
                  alt={selectedCountries[1].name}
                  width={60}
                  height={40}
                />

                <RemoveButton
                  onClick={() => removeCountry(selectedCountries[1])}
                />
              </div>
              <p>{selectedCountries[1].name}</p>
            </div>
          ) : (
            <div className="flex h-[40px] w-[60px] items-center justify-center self-center border">
              2
            </div>
          )}
        </div>

        {selectedCountries.length === 2 ? (
          <a
            href={`/country/compare/${selectedCountries[0].code}-${selectedCountries[1].code}`}
            className="rounded bg-[#1d1d1f] px-8 py-4 text-white"
          >
            See the comparison
          </a>
        ) : (
          <div className="">Select one more country to compare</div>
        )}
      </div>
    </div>
  );
}
