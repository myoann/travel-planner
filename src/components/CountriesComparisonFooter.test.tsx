import { render, screen } from "@testing-library/react";

import { TCountry } from "@/types";

import CountriesComparisonFooter from "./CountriesComparisonFooter";

const defaultProps = {
  selectedCountries: [
    {
      capital: "Paris",
      code: "FR",
      continent: {
        code: "EU",
        name: "Europe",
      },
      name: "France",
    } as TCountry,
    {
      capital: "Singapore",
      code: "SG",
      continent: {
        code: "AS",
        name: "Asia",
      },
      name: "Singapore",
    } as TCountry,
  ],
  removeCountry: jest.fn(),
};

describe("CountriesComparisonFooter", () => {
  it("matches the snapshot", () => {
    const { container } = render(
      <CountriesComparisonFooter {...defaultProps} />
    );

    expect(container).toMatchSnapshot();
  });

  it("displays the first selected country", () => {
    render(<CountriesComparisonFooter {...defaultProps} />);

    const image = screen.getByRole("img", { name: "France" });
    const name = screen.getByText("France");

    expect(image).toBeInTheDocument();
    expect(name).toBeInTheDocument();
  });

  it("displays the second selected country", () => {
    render(<CountriesComparisonFooter {...defaultProps} />);

    const image = screen.getByRole("img", { name: "Singapore" });
    const name = screen.getByText("Singapore");

    expect(image).toBeInTheDocument();
    expect(name).toBeInTheDocument();
  });

  it("displays a button to compare the selected countries", () => {
    render(<CountriesComparisonFooter {...defaultProps} />);

    const button = screen.getByRole("link", { name: "See the comparison" });

    expect(button).toBeInTheDocument();
  });

  it("displays a message if only one country is selected", () => {
    const selectedCountries = [defaultProps.selectedCountries[0]];
    render(
      <CountriesComparisonFooter
        selectedCountries={selectedCountries}
        removeCountry={defaultProps.removeCountry}
      />
    );

    const message = screen.getByText("Select one more country to compare");

    expect(message).toBeInTheDocument();
  });
});
