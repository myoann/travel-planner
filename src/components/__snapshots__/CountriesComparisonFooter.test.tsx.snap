// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountriesComparisonFooter matches the snapshot 1`] = `
<div>
  <div
    class="fixed bottom-0 w-full bg-white shadow-2xl z-50"
  >
    <p
      class="w-full flex items-center justify-center h-10 bg-[#1d1d1f] text-white font-bold"
    >
      Compare countries (
      2
      /2)
    </p>
    <div
      class="h-32 sm:h-52 bg-white flex items-center justify-center gap-16 sm:flex-col sm:gap-4"
    >
      <div
        class="flex text-center"
      >
        <div
          class="flex flex-col items-center w-32"
        >
          <div
            class="relative"
          >
            <img
              alt="France"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="/flags/fr.svg"
              style="color: transparent;"
              width="60"
            />
            <button
              class="absolute top-0 right-0 bg-white rounded-full"
              style="transform: translate(50%, -50%);"
            >
              <img
                alt="Remove image"
                data-nimg="1"
                decoding="async"
                height="16"
                loading="lazy"
                src="cross.svg"
                style="color: transparent;"
                width="16"
              />
            </button>
          </div>
          <p>
            France
          </p>
        </div>
        <span
          class="mx-4 text-4xl self-center"
        >
          +
        </span>
        <div
          class="flex flex-col items-center w-32"
        >
          <div
            class="relative"
          >
            <img
              alt="Singapore"
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="/flags/sg.svg"
              style="color: transparent;"
              width="60"
            />
            <button
              class="absolute top-0 right-0 bg-white rounded-full"
              style="transform: translate(50%, -50%);"
            >
              <img
                alt="Remove image"
                data-nimg="1"
                decoding="async"
                height="16"
                loading="lazy"
                src="cross.svg"
                style="color: transparent;"
                width="16"
              />
            </button>
          </div>
          <p>
            Singapore
          </p>
        </div>
      </div>
      <a
        class="bg-[#1d1d1f] text-white px-8 py-4 rounded"
        href="/country/compare/FR-SG"
      >
        See the comparison
      </a>
    </div>
  </div>
</div>
`;
